fury/actor.py

```diff
@@ -2452,3 +2452,73 @@ def texture_on_sphere(rgb, theta=60, phi=60, interpolate=True):
     earthActor.SetTexture(atext)
 
     return earthActor
+
+
+def sdf(centers, directions=(1, 0, 0), colors=(255, 0, 0), primitive='torus'):
+    """Create a SDF actor
+
+    Parameters
+    ----------
+    centers : ndarray, shape (N, 3)
+        SDF primitive positions
+    colors : ndarray (N,3) or (N, 4) or tuple (3,) or tuple (4,)
+        RGB or RGBA (for opacity) R, G, B and A should be at the range [0, 1]
+    directions : ndarray, shape (N, 3)
+        The orientation vector of the cube.
+    primitive : String
+        The primitive of choice to be rendered
+
+    Returns
+    -------
+    vtkActor
+    """
+
+    prims = {'sphere': 1.0, 'torus': 2.0}
+    verts, faces = fp.prim_box()
+
+    repeated = fp.repeat_primitive(verts, faces, centers=centers,
+                                   colors=colors, directions=directions)
+
+    rep_verts, rep_faces, rep_colors, rep_centers = repeated
+    box_actor = get_actor_from_primitive(rep_verts, rep_faces, rep_colors)
+
+    vtk_center = numpy_support.numpy_to_vtk(rep_centers)
+    vtk_center.SetNumberOfComponents(3)
+    vtk_center.SetName("center")
+    box_actor.GetMapper().GetInput().GetPointData().AddArray(vtk_center)
+
+    vs_dec_code = fs.load("sdf_dec.vert")
+    vs_impl_code = fs.load("sdf_impl.vert")
+    fs_dec_code = fs.load("sdf_dec.frag")
+    fs_impl_code = fs.load("sdf_impl.frag")
+
+    mapper = box_actor.GetMapper()
+    mapper.MapDataArrayToVertexAttribute(
+        "center", "center", vtk.vtkDataObject.FIELD_ASSOCIATION_POINTS, -1)
+
+    mapper.AddShaderReplacement(
+        vtk.vtkShader.Vertex, "//VTK::ValuePass::Dec", True,
+        vs_dec_code, False)
+
+    mapper.AddShaderReplacement(
+        vtk.vtkShader.Vertex, "//VTK::ValuePass::Impl", True,
+        vs_impl_code, False)
+
+    mapper.AddShaderReplacement(
+        vtk.vtkShader.Fragment, "//VTK::ValuePass::Dec", True,
+        fs_dec_code, False)
+
+    mapper.AddShaderReplacement(
+        vtk.vtkShader.Fragment, "//VTK::Light::Impl", True,
+        fs_impl_code, False)
+
+    @vtk.calldata_type(vtk.VTK_OBJECT)
+    def vtkShaderCallback(caller, event, calldata=None):
```

Comment:
I asked @skoudoro about this, he said for now to stick with shaderCallback.

Reply:
Same comment as above, let's talk about it in the next meeting

Reply:
@skoudoro @guaje If we use AddArray() the user has to define the primitive for those many cubes.
Eg: if I have 9 cubes, the user has to pass ['sphere', 'torus', ..., 'torus'], while in case of setUniform the user has to define the primitive only once, so what approach should I proceed with?